sycl/include/sycl/kernel_bundle.hpp

```diff
@@ -236,6 +238,19 @@ class kernel_bundle : public detail::kernel_bundle_plain,
     return kernel_bundle_plain::has_kernel(KernelID, Dev);
   }
 
+  /// \returns true only if the kernel bundle contains the kernel identified by
+  /// KernelName.
+  template <typename KernelName> bool has_kernel() const noexcept {
+    return has_kernel(get_kernel_id<KernelName>());
```

Comment:
@AlexeySachkov @steffenlarsen Should these return false if KernelName is not the name of a valid kernel? At the moment, get_kernel_id throws an exception if there isn't a kernel with this name. I think this is the same for the other two non-template has_kernel funtions, also?

Reply:
The specification states that
> Applications which call these functions for a KernelName that is not defined are ill formed, and the implementation must issue a diagnostic in this case.

which is also what it says about `get_kernel_id`, so I'm for letting it throw.

Reply:
We need to agree on the meaning of "diagnostic" here. `has_kernel` has `noexcept` attached to it, which makes me thing that the spec intended to see a compilation error rather than `std::terminate` at runtime. Tagging @Pennycook and @gmlueck for their input about meaning of "diagnostic"

Reply:
I was assuming these APIs would all trigger a link-time error if `KernelName` did not exist.

Reply:
Good point. I did not realize it was `noexcept`. We could potentially change `get_kernel_id` to fail at compile time if `getName()` is an empty string (default value if `KernelInfo` is unspecified). It may require some test changes, but it would probably be more helpful either way.

Reply:
Note that these APIs taking `KernelName` could reference a kernel that is defined in another translation unit.  Therefore, it seems like it would be hard to diagnose an invalid `KernelName` at compile-time.  That's why I was expecting a link-time error.

Reply:
I do not believe we are able to check this at link-time currently, so I suggest we go with a compile-time error for now then relax this limitation when we find a way to look the kernels up at link-time.

Reply:
I think it would be easy to diagnose this at link time.  How do we implement `get_kernel_id` now?  I was assuming we would do something like this in the integration header:

```
// Definition of "kernel_id" (this is just a dummy)
struct kernel_id {
  int i;
};

// Template function declaration for get_kernel_id() (not a definition).
template<typename KernelName> kernel_id get_kernel_id();

// For each kernel defined in this tranlsation unit, define a specialization
// of get_kernel_id
class Kernel1;
template<> kernel_id get_kernel_id<Kernel1>() {
  return {1};
}
```

User code in a translation unit can then call `get_kernel_id` for any kernel defined anywhere in the application (even in other translation units).  If there is no definition of a kernel, there will be no associated specialization of `get_kernel_id`, and that will provoke a link-time error.

Reply:
Currently our `kernel_id`s are created from kernel names, so the way `get_kernel_id` currently works is it gets the name associated with the kernel, as specified in the integration header, then finds the kernel ID associated with it. I do not believe that will continue to work if we expect link-time failure.

Reply:
Wouldn't the code pattern in the integration header be similar for getting the name of a kernel?

```
// Template function declaration for get_kernel_name() (not a definition).
template<typename KernelName> const char *get_kernel_name();

// For each kernel defined in this tranlsation unit, define a specialization
// of get_kernel_name
class Kernel1;
template<> kernel_id get_kernel_id<Kernel1>() {
  return "MyName";
}
```

Reply:
I guess handling with kernels coming from dynamically linked libraries is out of the scope of this discussion?

Reply:
I think these designs still work for the case when a kernel is defined in a shared library that is specified on the application's link line.  They do not work if the application loads a shared library dynamically via `dlopen`.  I think it's OK if we don't handle that.

Reply:
Currently the kernel name is a `constexpr` function inside the specialization of `KernelInfo`, where the default definition of `KernelInfo` has the function return an empty string. If we want the aforementioned link-time behavior I believe we will need to either rework the kernel name resolution or do as @gmlueck suggested in https://github.com/intel/llvm/pull/8070#discussion_r1088961081. This rework should not block these changes though, so I have opened a new issue for tracking it: https://github.com/intel/llvm/issues/8150